### PR TITLE
Add forecast components handling to base classes of models

### LINKED ISSUE #1124: Add forecast components handling to base classes of models
### 🚀 Feature Request

Add methods to obtain forecast components to base classes of models 

### Proposal

1. In classes `...AbstractModel`:
- Add flag `return_components: bool = False` to the signature of methods `forecast`, `predict`
2. In class `ModelForecastMixin`:
- Add methods:
```
def _forecast_components(self, **kwargs) -> pd.DataFrame:
def _predict_components(self, **kwargs) -> pd.DataFrame:
```
3. In classes `...ModelMixin`
- Add flag `return_components: bool = False` to the signature of methods `forecast`, `predict` 
- Implement the logic for `return_component=True` + merge the components to forecasts

4. In classes `PerSegmentModelMixin` and `MultiSegmentModelMixin`
- Implement methods `_forecast_components`/`_predict_components` calling the `forecast_components`/`predict_components` of base model
5. In `DeepBaseModel`, `DeepArModel`, `TFTModel`
- Add flag `return_components: bool = False` to the signature of methods `forecast`, `predict` and implement the corresponding logic raising `NotImplementedError`

### Test cases

- Test that `...ModelMixin` call `_forecast_components`/`_predict_components` in case of `return_components=True` in `forecast/predict`
- Test that `PerSegmentModelMixin`, `MultiSegmentModelMixin` call `forecast_components`/`predict_components` of adapter in `forecast_components`/`predict_components`
- Test that merge logic for forecast and forecast components works correctly

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by alex-hse-repository**
## Before submitting (must do checklist)

- [ ] Did you read the [contribution guide](https://github.com/tinkoff-ai/etna/blob/master/CONTRIBUTING.md)?
- [ ] Did you update the docs? We use Numpy format for all the methods and classes.
- [ ] Did you write any new necessary tests?
- [ ] Did you update the [CHANGELOG](https://github.com/tinkoff-ai/etna/blob/master/CHANGELOG.md)?

## Closing issues

closes #1124

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -7,6 +7,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## Unreleased
 ### Added
+- Target components logic base classes of models ([#1158](https://github.com/tinkoff-ai/etna/pull/1158))
```
Comment (Mr-Geekman) [inline CHANGELOG.md:10]: It seems like there is a missing preposition, like "into".

File: etna/models/mixins.py [hunk 2]
```diff
@@ -28,45 +28,65 @@ def _forecast(self, **kwargs) -> TSDataset:
...
+        return_components:
+            If True additionally returns forecast components
 
         Returns
         -------
         :
             Dataset with predictions
         """
-        return self._forecast(ts=ts)
+        forecast = self._forecast(ts=ts)
```
Comment (Mr-Geekman) [inline etna/models/mixins.py:58]: Doesn't we want to share this logic somehow?

File: etna/models/mixins.py [hunk 3]
```diff
@@ -348,6 +414,33 @@ def _make_predictions(self, ts: TSDataset, prediction_method: Callable, **kwargs
             ts.df = ts.df.iloc[-prediction_size:]
         return ts
 
+    def _make_component_predictions(self, ts: TSDataset, prediction_method: Callable, **kwargs) -> pd.DataFrame:
```
Comment (Mr-Geekman) [inline etna/models/mixins.py:417]: Why do we need this method? How is it different from `_make_predictions`?

File: etna/models/mixins.py [hunk 4]
```diff
@@ -418,6 +527,30 @@ def _make_predictions(self, ts: TSDataset, prediction_method: Callable, **kwargs
         ts.loc[:, pd.IndexSlice[:, "target"]] = y
         return ts
 
+    def _make_component_predictions(self, ts: TSDataset, prediction_method: Callable, **kwargs) -> pd.DataFrame:
```
Comment (Mr-Geekman) [inline etna/models/mixins.py:530]: Same question.

File: etna/models/mixins.py (hunk 3 above)
Comment (alex-hse-repository) [reply etna/models/mixins.py:417]: Here we create dataframe with components and in `_make_predictions` we create dataset with forecast. There is some difference in logic

File: tests/test_models/test_mixins.py [hunk 5]
```diff
@@ -124,3 +149,90 @@ def test_save_mixin_load_warning(get_version_mock, save_version, load_version, t
...
+    [
+        (PredictionIntervalContextIgnorantModelMixin, {}),
+        (NonPredictionIntervalContextIgnorantModelMixin, {}),
+        (PredictionIntervalContextRequiredModelMixin, {"prediction_size": 10}),
+        (NonPredictionIntervalContextRequiredModelMixin, {"prediction_size": 10}),
+    ],
+)
+@pytest.mark.parametrize("return_components", (True, False))
+def test_model_mixins_calls_add_target_components_in_forecast(mixin_constructor, return_components, call_params):
+    with patch.multiple(mixin_constructor, __abstractmethods__=set()):
```
Comment (Mr-Geekman) [inline tests/test_models/test_mixins.py:165]: What is it for? Looks very suspicious.

File: tests/test_models/test_mixins.py [hunk 6]
```diff
@@ -124,3 +149,90 @@ def test_save_mixin_load_warning(get_version_mock, save_version, load_version, t
...
+        mixin._add_target_components.assert_called_with(
+            ts=ts,
+            predictions=predict_ts,
+            components_prediction_method=mixin._predict_components,
+            return_components=return_components,
+        )
+
+
+@pytest.mark.parametrize("mixin_constructor", [PerSegmentModelMixin])
+def test_make_prediction_segment_with_components(mixin_constructor, target_components_df, ts_without_target_components):
```
Comment (Mr-Geekman) [inline tests/test_models/test_mixins.py:210]: We haven't changed `_make_predictions_segment` in this task, do we really need this test? 

Should we add other tests on this method as well (no components, handling `prediction_size`, etc)?

Comment (alex-hse-repository) [reply tests/test_models/test_mixins.py:210]: Here we check how it works with target components

File: tests/test_models/test_mixins.py [hunk 7]
```diff
@@ -5,18 +5,102 @@
...
+from etna.models.mixins import NonPredictionIntervalContextIgnorantModelMixin
+from etna.models.mixins import NonPredictionIntervalContextRequiredModelMixin
 from etna.models.mixins import PerSegmentModelMixin
+from etna.models.mixins import PredictionIntervalContextIgnorantModelMixin
+from etna.models.mixins import PredictionIntervalContextRequiredModelMixin
 from etna.models.mixins import SaveNNMixin
 
 
+class DummyPredictAdapter(BaseAdapter):
+    def fit(self, df: pd.DataFrame, **kwargs) -> "DummyAdapter":
```
Comment (Mr-Geekman) [inline tests/test_models/test_mixins.py:30]: Typing is incorrect

File: tests/test_models/test_mixins.py [hunk 8]
```diff
@@ -5,18 +5,102 @@
...
+        df["target_component_b"] = 180
+        df = df.drop(columns=["target"])
+        return df
+
+    def get_model(self) -> "DummyAdapter":
+        return self
+
+
+class DummyForecastPredictAdapter(DummyPredictAdapter):
+    def fit(self, df: pd.DataFrame, **kwargs) -> "DummyAdapter":
```
Comment (Mr-Geekman) [inline tests/test_models/test_mixins.py:48]: Typing is incorrect

**Review by Mr-Geekman — APPROVED** [review]

**Pull request merged**